app/actions/utils.ts

```diff
@@ -32,3 +33,90 @@ export const getStatusMessage = (status: AppNotificationStatus) => {
   }
   return null;
 };
+export const getKns = async (params: {
+  address: string;
+  contract: Contract;
+}): Promise<Domain> => {
+  const domain: any = {};
+
+  try {
+    const domainName = await params.contract.defaultNames(params.address);
+    // what do we do if this is false?
+    const isNameVerified =
+      (await params.contract.getDomainHolder(domainName)) === params.address;
+    if (!isNameVerified) null;
+    domain.name = `${domainName}.klima`;
+    const customImage = await params.contract.getDomainData(domainName);
+    if (customImage && JSON.parse(customImage).imgAddress) {
+      domain.image = JSON.parse(customImage).imgAddress;
+    } else {
+      const domains = await params.contract.domains(domainName);
+      const tokenId = domains.tokenId;
+      // if no image is in getDomainData get metadata from tokenURI and parse+decode it
+      const domainData = await params.contract.tokenURI(tokenId);
+      const domainDataDecoded = parseURL(domainData);
+      const domainMetadata = atob(domainDataDecoded.data);
+      // base64 decode metadata and get default image
+      const decodedDefaultImage = atob(
+        parseURL(JSON.parse(domainMetadata).image).data
+      );
+      domain.defaultImage = decodedDefaultImage;
+    }
+  } catch (error: any) {
+    console.log(error);
+  }
+  return domain;
+};
+export const getEns = async (params: { address: string }): Promise<Domain> => {
+  const ethProvider = ethers.getDefaultProvider(1);
+  const ens: any = {};
+  const ensDomain: string | null = await ethProvider.lookupAddress(
+    params.address
+  );
+  if (ensDomain) {
+    const avatar = ethProvider.getAvatar(ensDomain);
+    ens.avatar = avatar;
+  }
+  if (ensDomain) {
+    ens.name = `${ensDomain}.eth`;
+  }
+  return ens;
+};
+
+const parseURL = (url: string): any => {
```

Comment:
can you leave a what/why comment, i can't really  understand what this is doing lol

also lets avoid `any` return type, looks like you can make an interface for `parsed`

Reply:
updated